### PR TITLE
Add lockdown support

### PR DESCRIPTION
**Pull request by umutuzgur**
* Connect to lockdown from usbmux
* Get values from lockdownd
* Add the necessary encoder-decoders

### Comments
I still need to add getting the certs from usbmuxd and start a ssl connection with lockdown. Most of the lockdown apis are blocked a ssl session starts between lockdownd and the client

### CODE REVIEW

File: lib/lockdown/lockdown.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  async queryType () {
+    let data = await this.plistService.sendPlistAndReceive({
+      Label: 'usbmuxd',
+      ProtocolVersion: '2',
+      Request: 'QueryType'
+    });
+    if (data.Request === 'QueryType' && data.Type === 'com.apple.mobile.lockdown') {
+      return;
+    } else {
+      throw data;
```
Comment (imurchie) [inline lib/lockdown/lockdown.js]: Is `data` an `Error` in this case?

File: lib/lockdown/lockdown.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+class Lockdown {
+
+  constructor (plistService) {
+    this.plistService = plistService;
+  }
+
+  async queryType () {
+    let data = await this.plistService.sendPlistAndReceive({
+      Label: 'usbmuxd',
```
Comment (imurchie) [inline lib/lockdown/lockdown.js]: This and `ProtocolVersion` should probably be a constant.

File: lib/lockdown/lockdown.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
+const LOCKDOWN_PORT = 62078;
```
Comment (mykola-mokhnach) [inline lib/lockdown/lockdown.js]: is it always the same?

File: lib/lockdown/lockdown.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  async queryType () {
+    let data = await this.plistService.sendPlistAndReceive({
+      Label: LABEL,
+      ProtocolVersion: PROTOCOL_VERSION,
+      Request: 'QueryType'
+    });
+    if (data.Request === 'QueryType' && data.Type === 'com.apple.mobile.lockdown') {
+      return data;
+    } else {
```
Comment (mykola-mokhnach) [inline lib/lockdown/lockdown.js]: else is redundant

File: lib/lockdown/lockdown.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+const PROTOCOL_VERSION = 2;
+
+class Lockdown {
+
+  constructor (plistService) {
+    this.plistService = plistService;
+  }
+
+  async queryType () {
+    let data = await this.plistService.sendPlistAndReceive({
```
Comment (mykola-mokhnach) [inline lib/lockdown/lockdown.js]: const everywhere

File: lib/lockdown/lockdown.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      Service: serviceName,
+    });
+
+    if (data.Error) {
+      throw new Error(data);
+    } else {
+      return data;
+    }
+  }
+  close () {
```
Comment (mykola-mokhnach) [inline lib/lockdown/lockdown.js]: missing empty line

File: lib/plist-service/plist-service.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    this._splitter = new LengthBasedSplitter(false, 1000000, 0, 4, 4);
+    this._socketClient.pipe(this._splitter).pipe(this._decoder);
+
+    this._encoder = new PlistServiceEncoder();
+    this._encoder.pipe(this._socketClient);
+
+  }
+
+  async sendPlistAndReceive (json) {
+    this.sendPlist(json);
```
Comment (mykola-mokhnach) [inline lib/plist-service/plist-service.js]: await?

File: lib/plist-service/plist-service.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    return await this.receivePlist();
+  }
+
+  sendPlist (json) {
+    this._encoder.write(json);
+  }
+
+  async receivePlist () {
+    return await new B((resolve) => this._decoder.once('data', (data) => resolve(data)));
```
Comment (mykola-mokhnach) [inline lib/plist-service/plist-service.js]: is the promise always resolved or there might be a timeout?

File: lib/plist-service/plist-service.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  enableSessionSSL (hostPrivateKey, hostCertificate) {
+    this._socketClient.unpipe(this._splitter);
+    this._splitter.unpipe(this._decoder);
+    this._encoder.unpipe(this._socketClient);
+    this._socketClient = new tls.TLSSocket(this._socketClient, {
+      secureContext: tls.createSecureContext({
+        key: hostPrivateKey,
+        cert: hostCertificate,
+        rejectUnauthorized: false,
+        secureProtocol: 'TLSv1_method'
```
Comment (mykola-mokhnach) [inline lib/plist-service/plist-service.js]: this can be moved to a constant

File: lib/plist-service/plist-service.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      })
+    });
+    this._encoder.pipe(this._socketClient);
+    this._socketClient.pipe(this._splitter).pipe(this._decoder);
+  }
+  close () {
```
Comment (mykola-mokhnach) [inline lib/plist-service/plist-service.js]: empty line

File: lib/lockdown/lockdown.js (hunk 3 above)
Comment (umutuzgur) [reply lib/lockdown/lockdown.js]: Yeap. This is a daemon always running on the phone

File: lib/plist-service/plist-service.js (hunk 8 above)
Comment (umutuzgur) [reply lib/plist-service/plist-service.js]: That is actually a good point. I should also add this to the usbmux endpoints. What is the easiest way of setting that with bluebird?

Comment (imurchie) [reply lib/plist-service/plist-service.js]: http://bluebirdjs.com/docs/api/timeout.html

File: lib/lockdown/lockdown.js (hunk 4 above)
Comment (umutuzgur) [reply lib/lockdown/lockdown.js]: Should I remove that from every place that I do this if-else check?

**Pull request merged**